### ISSUE TITLE
Extend releases lock?

### ISSUE THREAD

**Issue by justinanastos** [mike-marcacci/node-redlock#11]
I have seen the following sequence of events and am wondering if I am using redlock incorrectly:

1. Function A requests a lock
1. Function B requests a lock (almost the same time)
1. Function A acquires lock.
1. Function A attempts to extend the lock.
1. Function B receives lock.
1. Function B attempts to extend the lock.
1. Deadlock.

Function A's attempt to extend the lock never resolves.

Is there possibly a bug in the extend code that causes the lock to be released instead of extended?

I am using only 1 redis instance.

**Comment by mike-marcacci** [comment]
Hi @justinanastos, sorry for the slow reply: just got back from some time off the grid.

This is interesting... the fact that **Function B** is able to receive the lock suggests that the extension attempt of **Function A** failed and its original lock expired. This doesn't sound like a true deadlock, but rather a callback that isn't being called or a promise that isn't being resolved.

Which API are you using for this – Promise, callback, or bluebird disposer?

**Comment by justinanastos** [comment]
I'm using the Promise API.

You're right, it's not an actual deadlock. Eventually both locks will throw a `LockError` for not being able to retrieve/extend the lock for `ttl`.

**Comment by justinanastos** [comment]
The problem, in way less words, is when Function B is waiting for a lock and Function A attempts to extend the lock it already has, things go wrong.

**Comment by justinanastos** [comment]
> This is interesting... the fact that Function B is able to receive the lock suggests that the extension attempt of Function A failed and its original lock expired

An error would be thrown if trying to extend an expired lock. I am not seeing that.

**Comment by mike-marcacci** [comment]
Hi @justinanastos, I finally got a moment to check out this repo again (actually looking for something else) and I added the following to the test, which succeeded under single and multi-redis conditions. Do these cover your use case? I'm not going to keep these in the test suite, as every step is already covered by more granular tests, but I wanted to make sure I wasn't missing something here.

```js
	it('should follow workflow from #11', function(done) {
		// acquire a lock
		redlock.lock(resource, 20000)
		.then(function(lockA) {
			// make sure it's exclusive
			return redlock.lock(resource, 20000)
			.then(function(lockB) {
				throw new Error('attempt B should not receive a lock');
			}, function(err) {
				// extend the lock
				return lockA.extend(10000)
				.then(function(lockA2) {
					// make sure it's still exclusive
					return redlock.lock(resource, 20000)
					.then(function(lockC) {
						throw new Error('attempt C should not receive a lock');
					}, function(err) {
						// release the lock
						return lockA2.unlock()
						.then(function() {
							// make sure it's free
							return redlock.lock(resource, 10000)
							.then(function(lockD) {
								return lockD.unlock();
							})
						})
					});
				})
			})
		})
		.done(done, done)
	});

	it('should follow workflow from #11 (race)', function(done) {
		var winner;

		Promise.all([
			redlock.lock(resource, 20000)
			.then(function(lockA){
				if (winner) throw new Error('Locked twice!!');
				winner = lockA;
				return lockA.extend(10000)
				.then(function(lockB){
					return lockB.unlock();
				})
			}, function(err) {
				return;
			})
		])
		.done(function() {
			done();
		}, done);
	});
```

**Comment by mike-marcacci** [comment]
The *only* case I can think of where something close to your described behavior might take place is under **extreme** contention for a single resource in a multi-redis environment:

Say you have 3 redis nodes, and you suddenly get many, many requests (A, B, and C) for the resource, `myresource`:
- Request A wins the race to Node 1
- Request B wins the race to Node 2
- Request C wins the race to Node 3

Since none of these requests received a majority, none of them are issued the lock; the losing votes are removed from each redis node, and the whole process is retried.

If the number of requests for `myresource` are so great that this scenario is able to play out repeatedly, then I could see timeouts when acquiring a lock. However, I would consider this is extremely unlikely, and this wouldn't effect extensions, which by definition don't race (their previously acquired locks are already exclusive).

If you still have the code that was experiencing this, can you post it? I don't really see a way this could be happening, but I don't want to dismiss something that could be a critical flaw.

**Comment by mike-marcacci** [comment]
One other thing I'd like to suggest in general is to listen for the `clientError` event and make sure these end up in your logs:

```js
redlock.on('clientError', function(err){ console.error(err); });
```

This event is emitted when there is a client library error that doesn't effect the intended result of a function (for example, failing to acquire a lock in 1 node of a 3 node setup). However, this can be extremely useful for diagnosing a redis issues.

**Comment by mike-marcacci** [comment]
Hey @justinanastos, it's actually tested and running in production on both versions. I might also mention that yesterday I was chasing down a bug in a system that uses this library. In this particular case `unlock` was failing because immediately prior to calling it, the worker was inserting a massive payload into a redis list (used as a queue) on the same single server, which triggered a `BGSAVE` that failed and caused the following error:

```
MISCONF Redis is configured to save RDB snapshots, but is currently not able to persist on disk. Commands that may modify the data set are disabled. Please check Redis logs for details about the error.
```

The simple, immediate fix was simply to update our redis configuration: `config set stop-writes-on-bgsave-error no`.

Redlock correctly and safely handled this case, and refused to reissue a lock on the same resource until it expired. While we were able to see this in the `clientError` events, I realized that because `unlock` failed silently, there was no ergonomic way to detect and respond to this inline.

Hence, I've changed `unlock` so that it now errors if it cannot confirm nonexistence of the lock in at least a majority of servers. This is technically a breaking change so I've just pushed v2.0.0.

To reiterate, this wasn't actually a failure in any way – just a less intuitive API – and doesn't explain your issues, but it may help you handle them better. You'll also definitely want to listen to the `clientError` in case there are errors on a minority of servers.
